src/main/kotlin/lotto/domain/Prize.kt

```diff
@@ -3,20 +3,25 @@ package lotto.domain
 enum class Prize(
     val value: Int,
     val matchCount: Int,
-    private val matcher: (Int, Boolean) -> Boolean,
 ) {
-    FIRST(2_000_000_000, 6, { count, _ -> count == 6 }),
-    SECOND(30_000_000, 5, { count, hasBonus -> count == 5 && hasBonus }),
-    THIRD(1_500_000, 5, { count, hasBonus -> count == 5 && !hasBonus }),
-    FOURTH(50_000, 4, { count, _ -> count == 4 }),
-    FIFTH(5_000, 3, { count, _ -> count == 3 }),
-    NONE(0, 0, { _, _ -> false }),
+    FIRST(2_000_000_000, 6),
+    SECOND(30_000_000, 5),
+    THIRD(1_500_000, 5),
+    FOURTH(50_000, 4),
+    FIFTH(5_000, 3),
+    NONE(0, 0),
     ;
 
     companion object {
         fun calculate(
             count: Int,
             hasBonus: Boolean,
-        ) = entries.find { it.matcher(count, hasBonus) } ?: NONE
+        ) = when (count) {
+            6 -> FIRST
+            5 -> if (hasBonus) SECOND else THIRD
+            4 -> FOURTH
+            3 -> FIFTH
+            else -> NONE
+        }
     }
```

Comment:
> I was thinking that the hasBonus is also part of the prize domain because it plays a significant role in differentiating SECOND and THIRD prize. However doing so is making the code more complicated with repetitions or unnecessary underscores. Even if I added a property with val requireBonus, the finding logic seems more difficult.
As a result, I refactored so that the prize is calculated on the calculate method. I feel like this has made the logic more simpler. Hope this was what you meant. 🙌

You get it - it's all about tradeoffs, but this seems more explicit to me 🙂